pkg/reconciler/certificate/certificate.go

```diff
@@ -117,8 +123,13 @@ func (c *Reconciler) reconcile(ctx context.Context, knCert *v1alpha1.Certificate
 	knCert.Status.ObservedGeneration = knCert.Generation
 
 	cmConfig := config.FromContext(ctx).CertManager
+	err := c.setHTTP01Challenges(knCert)
```

Comment:
I think we probably want to call `setHTTP01Challenges` after https://github.com/knative/serving/pull/5186/files?file-filters%5B%5D=.go#diff-e2259688f83ba9787c35d121c3cfd39bL138 because:
1. HTTP challenge resources do not exist before creating CM Certificate.
2. We don't need to call `setHTTP01Challenges` if the CM Certificate is already Ready, which is a little bit more efficient.

Reply:
If it gets moved to after the switch won't it still get called in the case of a ready certificate?

Reply:
You are right.

My point here is that for the ready certificate, we don't need to call `setHTTP01Challenges` to look up resources. In addition, we want to create CM certificate first before looking up HTTP challenge resources.

Reply:
In addition, if we use DNS challenge, we don't need to look up http01 challenge resources. We should add this condition check before calling setHTTP01Challenge.

Reply:
Yeah, that makes sense.